"""Coronavirus Variant analysis"""
import pkg_resources, os, pandas
from Bio import SeqIO
from Bio.Data.CodonTable import unambiguous_dna_by_id as codon_table
# path to data files internally used by cova
DATAPATH = pkg_resources.resource_filename('cova', 'data')
# feature table
FEATURETABLE =  pandas.read_csv( os.path.join( DATAPATH, 'feature_table_modified.txt'),
	sep='\t',index_col=0)
# proteins' sequence data
PSEQS = SeqIO.index( os.path.join( DATAPATH, 'protein_extended.faa'), 'fasta')
# CDS' sequence data
CDS = SeqIO.index( os.path.join( DATAPATH, 'cds_extended.fna'), 'fasta')
# reference genome sequence data
GENOME = SeqIO.read(handle = os.path.join( DATAPATH, 'genome.fna'), format='fasta')
# reference accession
REF = GENOME.id.split('.')[0]
# codon table
CODONTABLE = codon_table[1]
# proteins affected by ribosomal slippage
RFS = pandas.read_csv( os.path.join( DATAPATH, 'ribosomal_slippage.csv'), index_col=0)

# list of input/output file and directory names used and generated by CoVa
with open( os.path.join( DATAPATH, 'cova_io_files.txt')) as flob:
	IOF = [ i.strip('\n') for i in flob]

# data for sequence typing
with open(file=os.path.join( DATAPATH, 'barcode_pos.txt')) as flob:
	TYPEPOS = [ int(i.strip('\n')) for i in flob]

st = pandas.read_csv(os.path.join( DATAPATH, 'seq_types.csv'),index_col=0)
st = st.apply( axis = 1, func = lambda x: ''.join(x))     
SEQTYPES = dict(zip(st.values,st.index))  

# load internal modules
from ._func import rm_genome_w_stopm, extract_nmsa_prots,\
run_fubar, parse_fubar, genome_var, genome_seqtype
from . import _divs, _annotator
from . import utils
